FileUtilsTest fails on clean build

Trying to build Glassfish with JDK 21 yield the following error:

```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   FileUtilsTest.testCopyDirectoriesFiles:91 expected: <29> but was: <60>
[INFO] 
[ERROR] Tests run: 2116, Failures: 1, Errors: 0, Skipped: 2
```

```java
    @Test
    public void testCopyDirectoriesFiles() throws Exception {
        File outputDir = new File(tempDir, "outputDir");
        File testDir = new File(FileUtilsTest.class.getResource("/process").toURI());
        FileUtils.copy(testDir, outputDir);
        assertEquals(testDir.length(), outputDir.length());
    }
```


The value of both variables:

testDir: `/workspace/glassfish/nucleus/common/common-util/target/test-classes/process`
outputDir: `/tmp/junit4564666456115382502/outputDir`

The [Files::length](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/io/File.html#length()) states:  _The return value is unspecified if this pathname denotes a directory._

Is this test actually testing what it should? Is this test failure expected?

Forcing the tmp directory to be in the same file system (`-Djava.io.tmpdir=/workspace/my-tmp/`) works

```java
gitpod /workspace/glassfish (issue/25085) $ stat  /workspace/my-tmp/junit16177679752619825247/outputDir
  File: /workspace/my-tmp/junit16177679752619825247/outputDir
  Size: 29              Blocks: 0          IO Block: 4096   directory
Device: fd01h/64769d    Inode: 2088763522  Links: 2
Access: (0755/drwxr-xr-x)  Uid: (33333/  gitpod)   Gid: (33333/  gitpod)
Access: 2024-08-13 12:11:56.279965133 +0000
Modify: 2024-08-13 12:11:56.283965129 +0000
Change: 2024-08-13 12:11:56.283965129 +0000
 Birth: 2024-08-13 12:11:56.279965133 +0000
```

However, now other test fails 🫠

```
[ERROR] Tests run: 9, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 1.911 s <<< FAILURE! -- in org.glassfish.common.util.admin.locking.FileLockTest
[ERROR] org.glassfish.common.util.admin.locking.FileLockTest.lockAndRenameTest -- Time elapsed: 0.006 s <<< FAILURE!
org.opentest4j.AssertionFailedError: File renaming blocked ==> expected: <true> but was: <false>
        at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
        at org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
        at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
        at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:214)
        at org.glassfish.common.util.admin.locking.FileLockTest.lockAndRenameTest(FileLockTest.java:364)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

--- a/nucleus/common/common-util/src/main/java/com/sun/enterprise/util/io/FileUtils.java
+++ b/nucleus/common/common-util/src/main/java/com/sun/enterprise/util/io/FileUtils.java
@@ -40,6 +40,8 @@ import java.nio.channels.ReadableByteChannel;
 import java.nio.channels.WritableByteChannel;
 import java.nio.charset.Charset;
 import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
 import java.util.Collection;
 import java.util.Locale;
@@ -743,7 +745,7 @@ public final class FileUtils {
 
 
     /**
-     * Copies a file.
+     * Copies a file or directory.
      *
      * @param fin File to copy
      * @param fout New file
@@ -762,6 +764,34 @@ public final class FileUtils {
         }
         Files.copy(fin.toPath(), fout.toPath(), StandardCopyOption.REPLACE_EXISTING);
         LOG.log(DEBUG, "Successfully copyied file {0} to {1}", fin, fout);
+    }
+
+
+    /**
+     * Computes file or directory size. Follows symlinks just for the provided parameter, but not
+     * for files under the tree.
+     *
+     * @param fileOrDirectory
+     * @return summary of sizes of regular files.
+     * @throws IOException if any file size cannot be read.
+     */
+    public static long getSize(File fileOrDirectory) throws IOException {
+        try {
+            return Files.walk(fileOrDirectory.getCanonicalFile().toPath())
+                .filter(p -> Files.isRegularFile(p, LinkOption.NOFOLLOW_LINKS)).mapToLong(FileUtils::getRegularFileSize)
+                .sum();
+        } catch (IllegalStateException e) {
+            throw new IOException("Could not read size of " + fileOrDirectory, e);
+        }
+    }
+
+
+    private static long getRegularFileSize(Path path) {
+        try {
+            return Files.size(path);
+        } catch (IOException e) {
+            throw new IllegalStateException("Could not read file size for " + path, e);
+        }
     }
 
 
@@ -1057,7 +1087,7 @@ public final class FileUtils {
         private Throwable lastError;
         private final File out;
 
-        public FileOutputStreamWork(File out) {
+        private FileOutputStreamWork(File out) {
             this.out = out;
         }
 

--- a/nucleus/common/common-util/src/test/java/com/sun/enterprise/util/io/FileUtilsTest.java
+++ b/nucleus/common/common-util/src/test/java/com/sun/enterprise/util/io/FileUtilsTest.java
@@ -88,7 +88,7 @@ public class FileUtilsTest {
         File outputDir = new File(tempDir, "outputDir");
         File testDir = new File(FileUtilsTest.class.getResource("/process").toURI());
         FileUtils.copy(testDir, outputDir);
-        assertEquals(testDir.length(), outputDir.length());
+        assertEquals(FileUtils.getSize(testDir), FileUtils.getSize(outputDir));
     }
 
 

